### PR TITLE
MLDB-1632 path speedups 2

### PR DESCRIPTION
**Pull request by jeremybarnes**
More work on reducing the cost associated with structured paths as keys in the previous release.

### CODE REVIEW

**Comment by mailletf** [comment]
Is the failure intermitent?

@steadtler can you check this PR please?

**Comment by steadtler** [comment]
I doubt its intermitent. Waiting for the tests to pass to review it.

**Comment by mailletf** [comment]
@jeremybarnes back to you until tests pass

**Comment by jeremybarnes** [comment]
Test failures are mostly bugs in the tests (not asserting on the rowHash being present).  I'm working on them.

**Comment by jeremybarnes** [comment]
@mailletf @FinchPowers another review please; this should be good to go

File: sql/cell_value.cc [hunk 1]
```diff
@@ -34,9 +34,34 @@ namespace MLDB {
...
+        strFlags = 1;  // length of path
+
+        if (strLength <= INTERNAL_LENGTH) {
+            std::copy(p, p + strLength, shortString);
+            type = ST_SHORT_PATH;
+        }
+        else {
+            void * mem = malloc(sizeof(StringRepr) + strLength);
+            longString = new (mem) StringRepr;
+            std::copy(p, p + strLength, longString->repr);
```
Comment (FinchPowers) [inline sql/cell_value.cc]: Est-ce "memory safe"? Qu'est-ce qui arrive ici si le new StringRepr ou le std::copy lancent une exception?

File: sql/execution_pipeline_impl.cc [hunk 2]
```diff
@@ -672,57 +675,79 @@ doGetFunction(const Utf8String & functionName,
...
             };
 
         return { exec, leftRowName.resultInfo };
     }
 
+    if (functionName == "leftRowPath") {
+        auto leftRowPath
+            = left->doGetFunction("rowPath", args, leftFieldOffset(fieldOffset), argScope);
+        return leftRowPath;
+        return { leftRowPath.exec, leftRowPath.resultInfo };
```
Comment (FinchPowers) [inline sql/execution_pipeline_impl.cc]: Double return?

File: sql/path.cc [hunk 3]
```diff
@@ -1326,8 +1341,8 @@ operator == (const Path & other) const
     //if (digits_ != other.digits_)
     //    return false;
 
-    // Short circuit
-    if (offset(0) == 0 && other.offset(0) == 0) {
+    // Short circuit (currently offset(0) is always 0, so always taken).
+    if (/*offset(0) == 0 && other.offset(0) == 0*/ true) {
```
Comment (FinchPowers) [inline sql/path.cc]: Any reason to leave the if then?

File: sql/execution_pipeline_impl.cc [hunk 4]
```diff
@@ -672,57 +675,79 @@ doGetFunction(const Utf8String & functionName,
...
+        return { rightRowName.exec, rightRowName.resultInfo };
+    }
+
+    if (functionName == "rightRowPath") {
+        auto rightRowPath
+            = right->doGetFunction("rowPath", args, rightFieldOffset(fieldOffset), argScope);
+        return rightRowPath;
 
-        return { exec, rightRowName.resultInfo };
+        return { rightRowPath.exec, rightRowPath.resultInfo };
```
Comment (FinchPowers) [inline sql/execution_pipeline_impl.cc]: Double return. Ça fait quelque chose que j'ignore?

File: sql/cell_value.cc (hunk 1 above)
Comment (jeremybarnes) [inline sql/cell_value.cc]: std::copy is over a fundamental type, so it's noexcept
placement new with a noexcept constructor is noexcept.  I marked the StringRepr constructor noexcept to document that.

File: sql/execution_pipeline_impl.cc (hunk 2 above)
Comment (jeremybarnes) [inline sql/execution_pipeline_impl.cc]: Fixed.

**Comment by FinchPowers** [comment]
Commentaires mineurs. +1.

File: sql/path.cc (hunk 3 above)
Comment (jeremybarnes) [inline sql/path.cc]: Refactored.

File: sql/execution_pipeline_impl.cc (hunk 4 above)
Comment (jeremybarnes) [inline sql/execution_pipeline_impl.cc]: Fixed.

File: sql/cell_value.cc (hunk 1 above)
Comment (jeremybarnes) [inline sql/cell_value.cc]: I added a check of the return value of malloc (duh!) and comments about the noexceptness

**Pull request merged**